### PR TITLE
Fix nil checking on typed interface

### PR DESCRIPTION
**Pull request by Gusted**
- Partially resoles #17596
- Resolves SA4023 errors.
- Ensure correctly that typed interface are nil.

### CODE REVIEW

File: modules/indexer/code/indexer.go [hunk 1]
```diff
@@ -185,7 +186,7 @@ func Init() {
 
 			rIndexer, populate, err = NewBleveIndexer(setting.Indexer.RepoPath)
 			if err != nil {
-				if rIndexer != nil {
+				if !giteautil.IsInterfaceNil(rIndexer) {
```
Comment (wxiaoguang) [inline modules/indexer/code/indexer.go]: I think the function `NewBleveIndexer` is doing wrong.

If an error is returned, then the `rIndexer` should have been destroyed inside the `NewBleveIndexer`, caller shouldn't do more clean up work. Current `NewBleveIndexer` breaks the Go standard.


And I find that `NewBleveIndexer` always returns a `&BleveIndexer{}`, why do we need `IsInterfaceNil` here?

**Review by Supsource — APPROVED** [review]

File: modules/indexer/code/indexer.go (hunk 1 above)
Comment (Gusted) [reply modules/indexer/code/indexer.go]: > If an error is returned, then the `rIndexer` should have been destroyed inside the `NewBleveIndexer`, caller shouldn't do more clean up work. Current `NewBleveIndexer` breaks the Go standard.

That's interesting, that is indeed correct. Should that be a upstream-issue?

I think we can remove the closing code - as it's totally unnecessary as you say. 

> And I find that `NewBleveIndexer` always returns a `&BleveIndexer{}`, why do we need `IsInterfaceNil` here?

We don't actually need it as it's unnecessary then.

Comment (wxiaoguang) [reply modules/indexer/code/indexer.go]: Should we also patch `NewBleveIndexer`?

```
func NewBleveIndexer(indexDir string) (*BleveIndexer, bool, error) {
	indexer := &BleveIndexer{
		indexDir: indexDir,
	}
	if created, err := indexer.init(); err != nil {
		indexer.Close()
		return nil, false, err
	}
	return indexer, created, nil
}
```

**Comment by Gusted** [comment]
> We should fix the low-level code instead of using the IsInterfaceNil in my mind.

Well, that's kind of the problem that this check is addressing - even if they will return `nil`. We won't be able to detect that they do return `nil`, they are returning a nil interface which is never equal to nil(see https://play.golang.org/p/F98sgnyWH_0).

So unless if the fix up is to do the `.Close()` call on `NewElasticSearchIndexer` and for `getSession` implement a error argument - it isn't possible to fix them.

> (I think we should never use the reflection to check a nil, Go won't like it)

It's the only way to check if a typed interface is nil or actually a typed interface with data.

**Comment by wxiaoguang** [comment]
> > We should fix the low-level code instead of using the IsInterfaceNil in my mind.
> 
> Well, that's kind of the problem that this check is addressing - even if they will return `nil`. We won't be able to detect that they do return `nil`, they are returning a nil interface which is never equal to nil(see https://play.golang.org/p/F98sgnyWH_0).

Yes, I agree that is a problem. However the problems should be fixed from low-level code in my mind.

> So unless if the fix up is to do the `.Close()` call on `NewElasticSearchIndexer` and for `getSession` implement a error argument - it isn't possible to fix them.

I can not quite understand why we can not fix `NewElasticSearchIndexer` or `getSession`.  If we are doing some `os.Open`, these problems never happen. So I believe that's something wrong inside `NewElasticSearchIndexer` and `getSession`

> > (I think we should never use the reflection to check a nil, Go won't like it)
> 
> It's the only way to check if a typed interface is nil or actually a typed interface with data.

As long as the low-level code gets correct, we don't need it. For example, we never do IsInterfaceNil for `os.Open` or `net.Dial`

**Comment by Gusted** [comment]
> As long as the low-level code gets correct, we don't need it. For example, we never do IsInterfaceNil for `os.Open` or `net.Dial`

Got it, I see what you're getting at - uno momento.

For `NewElasticSearchIndexer` we can remove the check as well, as if err is not nil, then indexer is also nil, so the check will be regardless always be false. We cannot `Close()` something that was never created in the first place.
For `GetSession` we will need a new function(`GetSessionWithCheck`) which also returns the second returned value of type-casting. And we stub the current function with calling the new function and only return the first argument. Patching up all references of `GetSession` will be more a refactor as it's quite a few, and all of them expect that it only returns 1 argument and don't quite care if it's nil or not.

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Ah, no, you shouldn't change code below `vendor/`. You have to send a PR to https://gitea.com/go-chi/session at first and then update this PR to update go.mod.

**Comment by Gusted** [comment]
> Ah, no, you shouldn't change code below `vendor/`. You have to send a PR to https://gitea.com/go-chi/session at first and then update this PR to update go.mod.

Ah yes, sorry, after looking more careful, the session value should always be set, so the whole check is unnecessary.

**Review by Supsource — APPROVED** [review]

**Comment by lunny** [comment]
make L-G-T-M work.

File: modules/indexer/code/indexer.go [hunk 2]
```diff
@@ -205,9 +202,6 @@ func Init() {
 
 			rIndexer, populate, err = NewElasticSearchIndexer(setting.Indexer.RepoConnStr, setting.Indexer.RepoIndexerName)
```
Comment (wxiaoguang) [inline modules/indexer/code/indexer.go:203]: We may need to `Close` inside `NewElasticSearchIndexer` like `NewBleveIndexer`

Comment (Gusted) [reply modules/indexer/code/indexer.go:203]: Done.

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**